aep/general/0146/aep.md.j2

```diff
@@ -1,5 +1,97 @@
 # Generic fields
 
-**Note:** This AEP has not yet been adopted. See
-[this GitHub issue](https://github.com/aep-dev/aep.dev/issues/27) for more
-information.
+Most fields in any API, whether in a request, a resource, or a custom response,
+have a specific type or schema. This schema is part of the contract that
+developers write their code against.
+
+However, occasionally it is appropriate to have a generic or polymorphic field
+of some kind that can conform to multiple schemata, or even be entirely
+free-form.
+
+## Guidance
+
+While generic fields are generally rare, a API **may** introduce generic field
+where necessary. There are several approaches to this depending on how generic
+the field needs to be; in general, APIs **should** attempt to introduce the
+"least generic" approach that is able to satisfy the use case.
+
+For example, an API **should not** use a completely generic field (such as
+`google.protobuf.Struct` in protobuf APIs) when the value of the field must
+correspond to one of a known number of schemas. Instead, the API **should** use
+a `oneof` to represent the known schemas.
+
+### Generic fields in protobuf APIs
+
+#### Oneof
+
+A `oneof` **may** be used to introduce a type union: the user or API is able to
+specify one of the fields inside the `oneof`. Additionally, a `oneof` **may**
+be used with the same type (usually strings) to represent a semantic difference
+between the options.
+
+Because the individual fields in the `oneof` have different keys, a developer
+can programmatically determine which (if any) of the fields is populated.
+
+A `oneof` preserves the largest degree of type safety and semantic meaning for
+each option, and APIs **should** generally prefer them over other generic or
+polymorphic options when feasible. However, the `oneof` construct is ill-suited
+when there is a large (or unlimited) number of potential options, or when there
+is a large resource structure that would require a long series of "cascading
+oneofs".
+
+**Note:** Adding additional possible fields to an existing `oneof` is a
+non-breaking change, but moving existing fields into or out of a `oneof` is
+breaking (it creates a backwards-incompatible change in Go protobuf stubs).
+
+#### Maps
+
+Maps **may** be used in situations where many values _of the same type_ are
```

Comment:
What would you think about specifying that a `google.protobuf.Struct` is not a valid value type in a map? A string:Struct map is a cheat to create untyped JSON.

Reply:
Are there also potentially valid use cases for a `map<string, Struct>`?  I suspect the answer is "yes" and that therefore we can't disallow it.